content/testable-architecture.md

```diff
@@ -3,4 +3,62 @@ title: Testable Architecture
 ---
 
-Testable Architecture
\ No newline at end of file
+Automated tests and executable specifications can only work with software that has a testable architecture.
+A testable architecture enables fast and reliable tests that are easy to write, execute and maintain.
+Fast tests enable the developers to run them and obtain fast feedback on what they are building, without losing focus or flow.
+
+When designing for testability, make sure that your products and services are composed of loosely-coupled,
+well-encapsulated components or modules; following the principles of
+[SOLID design](https://en.wikipedia.org/wiki/SOLID_(object-oriented_design)).
+
+Decoupling your business logic from your infrastructure (the often slow and brittle components),
+makes it possible to test at different levels, with maximum confidence and minimum cost.
+
+* Ensure that you can run your tests underneath your UI. Do not depend solely on UI tests. They are slow, expensive, and hard to fix.
+* Running tests against a database makes tests slow. And to make the tests behave consistently each time we run them,
+we need to make sure the database is in the expected state for each test, so the tests don't interfere with one another.
+
+We want our tests to use some kind of in-memory stub implementation, instead of an actual database.
+From the perspective of the domain logic, the behavior looks exactly the same.
+To make sure that the stub works in the same way as the real thing, we need to have confidence, which we can obtain using [contract tests](https://martinfowler.com/bliki/IntegrationContractTest.html).
+
+We can think of our database and our stub as two things that can be plugged into the same socket, or **port**.
+The application containing the business logic doesn't need to know it's talking to a stub rather than the actual database; it's simply talking to this port
+and storing and retrieving data. We can run our tests using both the stub and the real implementation. This gives us confidence that the stub behaves like the real thing.
+
+The business logic is at the core of the application, completely decoupled from external devices and services. It doesn’t know anything about databases, message queues,
+and web services because we’ve isolated them through these ports. Often, these ports are connected to adapters that interact with the real implementation. There will be an adapter for the database,
+an adapter for the queue, and an adapter for the web service. There is also a port that defines how we can interact with our system, such as
+a web server which plugs into this port and displays a UI in a browser. All the IO tends to happen outside of these ports.
+
+This architectural patterns is called the **ports and adapters pattern** (or [hexagonal architecture](http://alistair.cockburn.us/Hexagonal+architecture)).
+This pattern enables you to hook your scenarios and unit tests at a lower level, while the **contract tests** give you the confidence to do that.
```

Comment:
I would remove the word 'the' before ports and adapter. Can't motivate more than I think it reads better.

I would replace 'This pattern' with 'It' and thus avoid the word This twice in two sentences next to each other.

I would use the word 'connect' instead of 'hook'